### PR TITLE
Juice 30 and juice 44 save and load functionality

### PR DESCRIPTION
**Pull request by SamanthaFisk-SPUAccount**
Save and load are now in a workable state.

UI buttons "Load" and "Save As" have functionality. Pressing them will open their respective file dialogs, where you can select a file.

The main change is the plugin file_system.rs. It uses bevy's state system to remember which state it's in, Loading, Saving, or Running, and calls the correct save or load functions depending on that. I cannot call them on normal schedules because they have direct read/write access to the world. The actual saving and loading use bevy_save, which helps with a Pipeline and custom file formats (which I'm just using for the file extension right now). The file dialogs are made using rfd (Rust File Dialogs) library.

The actual file format is truly just .json, but I've renamed the file extension to .juice. Can be changed to something else if you guys want. Right now I have my test files in /assets/scenes. In /load_test are some older files that I made by manually editing them. Most of the files in save_test are not edited, except for the "oh no im trapped.juice" file.

Some things that are currently not working:
- Regular "Save" does not work because file_system doesn't save the current file right now.
- Cannot currently invoke a direct call to init_loading() with a preset filepath, so we can't set a default file or last used file right on startup. I can probably just save the default and most recent files in a metadata file somewhere, but relying on that would make the function less flexible.
- When the particles hit the ground I get this error: "WARN bevy_ecs::world: error[B0003]: Could not despawn entity 319v1 because it doesn't exist in this World." But it doesn't seem to be actually affecting anything at all as far as I can tell.
- The particles move strangely on the load of a file, to varying degrees.
- In some of the save files the velocities in SimGrid seem to be off. Insane numbers like -3.4028235e38 while some are 0.0019996716. Go to line 63732 in "interger overflow in entity count.juice" for an example. This is likely what's causing some funky moves of the particles
- Also, in that file the entities start count at 4294967299. I have no idea why or how to replicate it.

None of this bugginess that I've found has an impact on the performance of the first startup of the simulation at all, it only comes up when you try to load a file.

### CODE REVIEW

**Comment by r3dc0g** [comment]
Tested and works well! Thanks Matthew! 

Though, I don't think we need to keep test save files in the repo. If you want to add that folder to the git ignore, you could still work with them. I also think your vscode settings folder can be added to the git ignore. 

What's up with test.scn.ron ?

File: assets/scenes/load_test/test_load_11_friends.juice
Comment (Dillon-Goicoechea) [inline assets/scenes/load_test/test_load_11_friends.juice:1]: We should git ignore all non-UI assets.

File: src/file_system.rs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        app.add_state::<JuiceStates>();
+
+        // Setting up the type registry so the data can be accessed
+        app.register_type::<SimParticle>();
+        app.register_type::<Option<Vec2>>();
+        app.register_type::<Option<Rect>>();
+        app.register_type::<SimConstraints>();
+        app.register_type::<SimGrid>();
+        app.register_type::<SimGridCellType>();
+        app.register_type::<(u16, u16)>();
```
Comment (Dillon-Goicoechea) [inline src/file_system.rs]: what is this for?

File: src/file_system.rs [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        app.register_type::<Vec<Vec<SimGridCellType>>>();
+        app.register_type::<Vec<SimGridCellType>>();
+        app.register_type::<Vec<Vec<f32>>>();
+        app.register_type::<Vec<f32>>();
```
Comment (Dillon-Goicoechea) [inline src/file_system.rs:41]: please add comments for non explicitly named objects

File: src/file_system.rs [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        app.register_type::<Vec<Vec<Entity>>>();
+        app.register_type::<Vec<Entity>>();
+
+        app.add_systems(OnEnter(JuiceStates::Loading), load_scene);
```
Comment (Dillon-Goicoechea) [inline src/file_system.rs:55]: what are these doing?

File: src/file_system.rs [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    key: String,
+}
+
+impl JuicePipeline {
+    pub fn new(key: String) -> Self {
+        Self { key: key }
+    }
+}
+
+impl Pipeline for JuicePipeline {
```
Comment (Dillon-Goicoechea) [inline src/file_system.rs:128]: Please add comments explaining what each function is doing explicitly

File: src/file_system.rs [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+    file_state.set(JuiceStates::Saving); // Triggers save_scene()
+}
+
+/// Triggers a file dialog asking user for filepath, saves the data into the file. Function runs when state = JuiceStates::Saving.
+/// Does nothing if user doesn't select a file.
+fn save_scene(world: &mut World) {
+    let key = create_new_file(); // TODO: Run this in init_saving() and find a way to pass it into here.
+
+    if (key.is_some()) {
```
Comment (Dillon-Goicoechea) [inline src/file_system.rs]: is there a more clear or effective form of verification?

File: src/file_system.rs [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+        world
+            .save(JuicePipeline::new(key.unwrap()))
+            .expect("Did not save correctly, perhaps filepath was incorrect?");
+    }
+}
+
+/// Sets file_system.rs state to Loading, which triggers load_scene() to run.
+/// UNFINISHED FUNCTIONALITY - If a String is passed in the key argument, load that function. Otherwise run a file dialog asking the user.
```
Comment (Dillon-Goicoechea) [inline src/file_system.rs:202]: what part is unfinished?

File: src/file_system.rs [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+    }
+
+    file_state.set(JuiceStates::Loading); // Triggers load_scene()
+}
+
+/// Runs file dialog asking user for filepath, loads the file into the world. Function runs when state = JuiceStates::Loading.
+fn load_scene(world: &mut World) {
+    let key = get_file(); // TODO: Run this in init_loading() and find a way to pass it into here.
+
+    if (key.is_some()) {
```
Comment (Dillon-Goicoechea) [inline src/file_system.rs]: same concern as above

File: src/file_system.rs [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+    let start_path = std::env::current_dir().unwrap();
+
+    let selected_path = rfd::FileDialog::new()
+        .add_filter("text", &["juice"]) // Only allowing to select .json
+        .set_directory(&start_path) // Setting the initial folder of the file dialog to the assets folder
+        .pick_files();
+
+    if (selected_path.is_some()) {
+        let key: &mut String = &mut selected_path
+            .unwrap()[0]
```
Comment (Dillon-Goicoechea) [inline src/file_system.rs]: if any error is encountered this will crash the program

File: src/file_system.rs [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+/// Triggers a file dialog asking user to select an existing .juice file. Returns the path to it as an Option<String>.
+fn get_file() -> Option<String> {
+    let start_path = std::env::current_dir().unwrap();
+
+    let selected_path = rfd::FileDialog::new()
+        .add_filter("text", &["juice"]) // Only allowing to select .json
+        .set_directory(&start_path) // Setting the initial folder of the file dialog to the assets folder
+        .pick_files();
+
+    if (selected_path.is_some()) {
```
Comment (Dillon-Goicoechea) [inline src/file_system.rs]: same concern as above

File: src/file_system.rs [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+        let key: &mut String = &mut selected_path
+            .unwrap()[0]
+            .clone() // Allows us to move PathBuf since it can't be copied on it's own
+            .into_os_string()
+            .into_string()
+            .expect("Wasn't able to parse filepath");
+
+        key.truncate(key.len() - 6); // Removing the .juice file extension, bevy_save breaks otherwise.
```
Comment (Dillon-Goicoechea) [inline src/file_system.rs]: Can we please verify the key exists before attempting truncation?

File: src/file_system.rs [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+    let start_path = std::env::current_dir().unwrap();
+
+    let selected_path = rfd::FileDialog::new()
+        .add_filter("text", &["juice"])
+        .set_directory(&start_path)
+        .save_file();
+
+    if (selected_path.is_some()) {
+        let key: &mut String = &mut selected_path
+            .unwrap()
```
Comment (Dillon-Goicoechea) [inline src/file_system.rs]: same issue as before

File: src/file_system.rs [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+            .into_os_string()
+            .into_string()
+            .expect("Wasn't able to parse filepath");
+
+        key.truncate(key.len() - 6); // Removing the .juice file extension, bevy_save breaks otherwise.
```
Comment (Dillon-Goicoechea) [inline src/file_system.rs]: same issue as before

File: src/test.scn.ron
Comment (Dillon-Goicoechea) [inline src/test.scn.ron:1]: Please remove

**Review by Dillon-Goicoechea — CHANGES_REQUESTED** [review]
Looks like it is going in the right direction!

**Comment by kadealicious** [comment]
- Pls remove all unwrap() and expect() calls; a bad filepath/save file should not crash the program.
- There are a few sections with repeated code; see create_new_file() and get_file().  Putting some repeated code within smaller functions to increase readability is much preferred.
- // Comments comments comments comments comments comments pls
- /* Super long multi line comments are even preferred over no comments (We don't know how saving/loading works within this system, so it would be really hard for anyone else to make additions/upgrades/changes to the system without comments) */

File: .vscode/launch.json
Comment (r3dc0g) [inline .vscode/launch.json:1]: please put .vscode in .gitignore

File: assets/scenes/load_test/test_load_no_particles_all_solid.juice
Comment (r3dc0g) [inline assets/scenes/load_test/test_load_no_particles_all_solid.juice:1]: Also add testing save files to .gitignore

File: src/test.scn.ron
Comment (r3dc0g) [inline src/test.scn.ron:1]: Remove this file

**Review by r3dc0g — CHANGES_REQUESTED** [review]
Thanks, Matthew! I added my last requests for changes as some were missed.

**Review by r3dc0g — DISMISSED** [review]

**Review by Dillon-Goicoechea — APPROVED** [review]
lgtm

**Pull request merged**